model/syncQueue/storage/SyncQueueStorageRds.php

```diff
@@ -45,32 +46,99 @@ public function __construct($options = array())
+                ->setParameter('sync_type', $types, Connection::PARAM_STR_ARRAY);
+        }
+
+        $query
+            ->groupBy([self::PARAM_SYNCHRONIZABLE_ID, self::PARAM_EVENT_TYPE, self::PARAM_SYNC_MIGRATION_ID])
+            ->setMaxResults($limit);
+
+        return $query->execute()->fetchAll();
     }
 
-    public function getAll($limit = 1000, $offset = 0)
+    /**
+     * @param array $types
+     * @param int $limit
+     * @return array|mixed[]
+     */
+    public function getAggregatedQueued(array $types = [], $limit = 5000)
     {
-        $sql = 'SELECT * FROM ' . self::TABLE_NAME . ' ORDER BY ' . self::PARAM_ID . ' LIMIT ? OFFSET ?';
-        $parameters = [0];
-        $stmt = $this->getPersistence()->query($sql, $parameters);
-        return $stmt->fetchAll(PDO::FETCH_ASSOC);
+        return $this->getMigrationData(0, $types, $limit);
+    }
+
+    public function getAll($limit = 10000, $offset = 0)
+    {
+        $query = $this->getQueryBuilder()
+            ->select('*')
+            ->from(self::TABLE_NAME)
+            ->orderBy(self::PARAM_CREATED_AT)
+            ->setMaxResults($limit);
+
+        return $query->execute()->fetchAll();
+    }
+
+    /**
+     * Returns the QueryBuilder
+     *
+     * @return QueryBuilder
+     */
+    private function getQueryBuilder()
+    {
+        return $this->getPersistence()->getPlatform()->getQueryBuilder();
     }
 
     public function insert(array $action)
     {
         return $this->getPersistence()->insert(self::TABLE_NAME, $action);
     }
 
-    public function setSyncId($id)
+    public function setMigrationId($migrationId, $queuedTasks = [])
+    {
+        foreach ($queuedTasks as $queuedTask) {
+            $qb = $this->getPersistence()->getPlatForm()->getQueryBuilder();
+            $qb->update(static::TABLE_NAME)
+                ->set(static::PARAM_SYNC_MIGRATION_ID, ':migrationId')
+                ->set(static::PARAM_UPDATED_AT, ':date')
+                ->andWhere(static::PARAM_SYNCHRONIZABLE_ID . ' =:syncId')
+                ->andWhere(static::PARAM_EVENT_TYPE . ' =:eventType')
+                ->setParameter('migrationId', $migrationId)
+                ->setParameter('date', date('Y-m-d H:i:s'))
+                ->setParameter('syncId', $queuedTask[self::PARAM_SYNCHRONIZABLE_ID])
+                ->setParameter('eventType', $queuedTask[self::PARAM_EVENT_TYPE])
+                ->execute();
+        }
+        return true;
+    }
+
+    public function isSynchronized($eventType = '', $synchronizedIds = [])
     {
-        $sql = 'UPDATE ' . self::TABLE_NAME . ' SET ' . self::PARAM_SYNC_ID . ' = ?, '.self::PARAM_UPDATED_AT.'= ? WHERE ' . self::PARAM_ID . '= ?';
-        $parameters = [1, date('Y-m-d H:i:s'), $id];
-        return $this->getPersistence()->exec($sql, $parameters);
+        $query = $this->getQueryBuilder()
+            ->select('*')
+            ->from(self::TABLE_NAME)
+            ->andWhere(self::PARAM_EVENT_TYPE.'=:eventType')
```

Comment:
Since `$eventType` and `$synchronizedIds` are optional parameters I think it would be better to add these conditions in query builder only if value is provided. In current implementation if at least one parameter is not provided query will always return 0 rows.
If we add these conditions, it will be possible to check if there are data to synchronize by calling method without parameters.

Reply:
this method was improved, but we can't use it without params, I made them required